### PR TITLE
linter/markdown: adds support for languatool

### PR DESCRIPTION
**Pull request by wahrwolf**
As requested in #1516 and #2154, this will add support for the super awesome grammar/style checker [languagetool](https://languagetool.org/)

I provide this without any tests

Let me know if you have any feedback!

### CODE REVIEW

File: ale_linters/markdown/languagetool.vim [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	" Extract all messages
+	let l:message_matches = []
+	call extend(l:message_matches, ale#util#GetMatches(a:lines, l:message_pattern))
+
+
+	let i = 0
+	" Okay tbh I was to lazy to figure out a smarter solution here
+	" We just assume that the arrays are same sized and merge everything
+	" togther
```
Comment (oblitum) [inline ale_linters/markdown/languagetool.vim]: minor typo

Comment (wahrwolf) [reply ale_linters/markdown/languagetool.vim]: This is actually the reason I decided to write the linter plugin ;)

**Comment by oblitum** [comment]
The propaganda is so strong on those issues that I'm trying it now.

**Comment by oblitum** [comment]
In fact, I used LanguageTool from https://github.com/rhysd/vim-grammarous and forgot it was LanguageTool based :D

**Comment by wahrwolf** [comment]
Can you give me an example file, so I can try to reproduce the issue?

**Comment by oblitum** [comment]
I've opened a markdown file with the contents of example in issue #1516: https://github.com/w0rp/ale/issues/1516#issuecomment-383280147

**Comment by oblitum** [comment]
I get 8 errors, but ALE just highlighted the last error line in the file, all errors are in that line, which is wrong. Clocking on location list errors, it goes to the right column an error is, but doesn't jump to the affected line.

**Comment by wahrwolf** [comment]
Are you sure, that you inserted newline?
With my copy from the text I got no line breaks at all!

**Comment by oblitum** [comment]
Yes, I formatted with `gqg`.

**Comment by wahrwolf** [comment]
That is weird. I can reproduce the issue, but when I add newlines with `s/\./.\r/g` everything seems normal...

**Comment by wahrwolf** [comment]
This looks like an issues with languagetool. It reports issues in lines that are not even on the screen...

File: ale_linters/markdown/languagetool.vim [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        let l:i+=1
+    endwhile
+
+    return l:output
+endfunction
+
+call ale#linter#Define('markdown', {
+            \   'name': 'languagetool',
+            \   'executable': 'languagetool',
+            \   'command': 'languagetool --line-by-line %s ',
```
Comment (skywind3000) [inline ale_linters/markdown/languagetool.vim]: would parameter `--line-by-line` break setences acrossing lines ?

**Comment by oblitum** [comment]
```
> cat foo.md

Paste your own text here and click the 'Check Text' button. Click the colored
phrases for details on potential errors. or use this text too see an few of of
the problems that LanguageTool can detecd. What do you thinks of grammar
checkers?lease not that they are not perfect.Style issues get a blue marker:
It's 5.M. in the afternoon. LanguageTool 3.8 was released on Thursday, 27 June
2017.

> languagetool foo.md

1.) Line 2, column 42, Rule ID: UPPERCASE_SENTENCE_START
Message: This sentence does not start with an uppercase letter
Suggestion: Or
...red phrases for details on potential errors. or use this text too see an few of of the probl...
                                                ^^                                             

2.) Line 2, column 59, Rule ID: TOO_TO[1]
Message: Did you mean 'to see'?
Suggestion: to see
...etails on potential errors. or use this text too see an few of of the problems that LanguageTool ...
                                                ^^^^^^^                                             

3.) Line 2, column 67, Rule ID: EN_A_VS_AN
Message: Use 'a' instead of 'an' if the following word doesn't start with a vowel sound, e.g. 'a sentence', 'a university'
Suggestion: a
...n potential errors. or use this text too see an few of of the problems that LanguageTool can...
                                                ^^                                             

4.) Line 2, column 74, Rule ID: ENGLISH_WORD_REPEAT_RULE
Message: Possible typo: you repeated a word
Suggestion: of
...tial errors. or use this text too see an few of of the problems that LanguageTool can detecd. W...
                                                ^^^^^                                             

5.) Line 3, column 56, Rule ID: DO_VBZ[1]
Message: Consider using the base form 'think'.
Suggestion: think
...ms that LanguageTool can detecd. What do you thinks of grammar checkers?lease not that they are ...
                                                ^^^^^^                                             

6.) Line 3, column 56, Rule ID: NON3PRS_VERB[4]
Message: The pronoun 'you' must be used with a non-third-person form of a verb: 'think'
Suggestion: think
...ms that LanguageTool can detecd. What do you thinks of grammar checkers?lease not that they are ...
                                                ^^^^^^                                             

7.) Line 4, column 46, Rule ID: SENTENCE_WHITESPACE
Message: Add a space between sentences
Suggestion:  Style
...checkers?lease not that they are not perfect.Style issues get a blue marker: It's 5.M. in the a...
                                                ^^^^^                                             

8.) Line 5, column 8, Rule ID: SENTENCE_WHITESPACE
Message: Add a space between sentences
Suggestion:  M
...rfect.Style issues get a blue marker: It's 5.M. in the afternoon. LanguageTool 3.8 was rele...
                                                ^                                             

9.) Line 5, column 62, Rule ID: DATE_WEEKDAY[1]
Message: The date 27 June 2017 is not a Thursday, but a Tuesday.
... afternoon. LanguageTool 3.8 was released on Thursday, 27 June 2017. 
                                                ^^^^^^^^^^^^^^^^^^^^^^  
Time: 1713ms for 7 sentences (4.1 sentences/sec)
```

**Comment by oblitum** [comment]
Exactly, the `line-by-line` parameter cause it to output strange lines, without it seems fine.

**Comment by oblitum** [comment]
Everything gets correctly highlighted without the parameter now removed from pull.

**Comment by oblitum** [comment]
I wished ALE linters were better at highlighting words....

**Comment by skywind3000** [comment]
I am using https://github.com/rhysd/vim-grammarous for LanguageTool right now.

**Comment by wahrwolf** [comment]
Okay I'll remove the param!

**Comment by oblitum** [comment]
It's a bit of extra work, but, the tool does highlight entire pieces of text while this addition just highlights a single char, so, could that be added? If not, I think it's to be born with an issue open on that.

Related: #2125

**Comment by oblitum** [comment]
Possibly just counting the amount of ^^^^^^ in the empty line and using it as the number of chars to highlight will do a good job.

**Comment by oblitum** [comment]
> Because most compilers or linters only output a position instead of a text range ?

I think vim-grammarous simply cares to process what I requested above. See #2125.

**Comment by wahrwolf** [comment]
alright, should work just fine!

**Comment by oblitum** [comment]
I've never seen a commit stream with that many typos, are you fine @wahrwolf?

**Comment by wahrwolf** [comment]
I am currently working on stuff for my university and am little low on sleep ;)

File: ale_linters/markdown/languagetool.vim [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            \}
+        call add(l:output, l:item)
+        let l:i+=1
+    endwhile
+
+    return l:output
+endfunction
+
+call ale#linter#Define('markdown', {
+    \   'name': 'languagetool',
```
Comment (w0rp) [inline ale_linters/markdown/languagetool.vim]: Put `\` on the same level of indentation as the preceding lines. I think I'll improve the syntax checking script so it spots this.

File: ale_linters/markdown/languagetool.vim [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    let l:output = []
+
+    " Extract the header line first
+    let l:head_matches = []
+    call extend(l:head_matches, ale#util#GetMatches(a:lines, l:head_pattern))
+
+    " Extract all messages
+    let l:message_matches = []
+    call extend(l:message_matches, ale#util#GetMatches(a:lines, l:message_pattern))
```
Comment (w0rp) [inline ale_linters/markdown/languagetool.vim]: You could write `let l:message_matches = ale#util#GetMatches(a:lines, l:message_pattern)` and so on. Either way is fine.

File: ale_linters/markdown/languagetool.vim (hunk 2 above)
Comment (wahrwolf) [reply ale_linters/markdown/languagetool.vim]: good point! tbh I don't know...
But it seems to work fine if the param is not set, so we just removing it...

**Comment by skywind3000** [comment]
@wahrwolf , thanks for your effort, it appears that this PR is only for markdown ?? can I use languagetool for plain text files ??

**Comment by wahrwolf** [comment]
@skywind3000 good point! let me write the handler for txt and mail too!

**Comment by oblitum** [comment]
bump.

**Comment by skywind3000** [comment]
I have made linters for ALE too, everything worked fine. But it required tests and I encountered some vader (?? If I remember correctly)  integration issues on my laptop, so I give up.

the corpse of my PR is still there:
https://github.com/w0rp/ale/pull/1571

The hardest part is not writing the linter or tests, but how to solve problem in vader.

**Comment by w0rp** [comment]
See `:help ale-development-linter-tests` for information on adding the tests required for linter commands and executables.

**Comment by wahrwolf** [comment]
Oh i forgotten completly this Ticket... Going to fix it this week though.
Ty for the reminder

Am So., 27. Jan. 2019, 16:59 hat w0rp <notifications@github.com>
geschrieben:

**Comment by noahhdf** [comment]
> let me write the handler for txt and mail too!

Is it possible (probably with more work) to use this with Latex files?

**Comment by oblitum** [comment]
> I have made linters for ALE too, everything worked fine. But it required tests and I encountered some vader (?? If I remember correctly) integration problems on my laptop, so I give up.
> 
> the corpse of my PR is still there:
> #1571
> 
> The hardest part is not writing the linter, but using vader.

Heh, I just wrote a very simple linter for Neo4j's Cypher, which is working by the way. I wished to contribute but the tests part of the process look quite annoying for someone not familiar with neither Vader nor the ALE specific test setup.

File: autoload/ale/handlers/languagetool.vim [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    " Okay tbh I was to lazy to figure out a smarter solution here
+    " We just check that the arrays are same sized and merge everything
+    " together
+    let l:i = 0
+
+    while l:i < len(l:head_matches) &&
+    \   (
+    \       (len(l:head_matches) == len(l:markers_matches)) &&
```
Comment (w0rp) [inline autoload/ale/handlers/languagetool.vim]: Move the binary operators to the start of the following line. The linting tools will complain about this now.

**Comment by oblitum** [comment]
bump 😴

**Comment by skywind3000** [comment]
@wahrwolf , will you make this final move ??

**Comment by wahrwolf** [comment]
now on to it! (i have the night and some coffee ready to go ;D )

**Comment by wahrwolf** [comment]
@w0rp lets try this :D

File: autoload/ale/handlers/languagetool.vim (hunk 5 above)
Comment (wahrwolf) [reply autoload/ale/handlers/languagetool.vim]: Done!

**Comment by w0rp** [comment]
Cheers! :beers:

**Comment by oblitum** [comment]
:rotating_light: :rotating_light: :rotating_light:

This broke mail/markdown/text/etc when languagetool is present. Sadly it's broken. On file write I get some java GUIs popping up due to `--autoDetect` not being recognized as a file, this happens when you run `languangetool --autoDetect` without a file parameter, this means the pull is not passing it, there's no `%s` in the `command_callback`.

**Comment by oblitum** [comment]
I think is the only issue, the missing `%s`.

**Comment by fszymanski** [comment]
Maybe the `languagetool_executable` should be set to something like this: `java -jar /path/to/languagetool-commandline.jar`?

**Comment by oblitum** [comment]
@fszymanski why? I see no reason for that.

**Comment by oblitum** [comment]
It's configurable. I think `languagetool` as default is expected.

**Comment by skywind3000** [comment]
Cannot be used on Windows:

```VimL
let ale_linters.markdown = ['languagetool']
let g:ale_languagetool_executable = 'd:/software/LanguageTool/LanguageTool-4.4/languagetool.cmd'
```

while `languagetool.cmd` is a batch file for launching `languagetool-commandline.jar`:

```
@ECHO off
set "INSTALLDIR=%~dp0"
java -jar "%INSTALLDIR%languagetool-commandline.jar" %*
```

I can see the java process hangs in the background, and ALEInfo returns:

aleerror

**Comment by oblitum** [comment]
As I mentioned previously, there's a bug even on Linux, the file is not being passed, there's a `%s` missing in the pull. This bug has nothing to do with the executable path.

```diff
diff --git a/autoload/ale/handlers/languagetool.vim b/autoload/ale/handlers/languagetool.vim
index aaad4c9..10e049d 100644
--- a/autoload/ale/handlers/languagetool.vim
+++ b/autoload/ale/handlers/languagetool.vim
@@ -10,7 +10,7 @@ endfunction
 function! ale#handlers#languagetool#GetCommand(buffer) abort
     let l:executable = ale#handlers#languagetool#GetExecutable(a:buffer)

-    return ale#Escape(l:executable) . ' --autoDetect '
+    return ale#Escape(l:executable) . ' --autoDetect %s'
 endfunction

 function! ale#handlers#languagetool#HandleOutput(buffer, lines) abort
```

**Comment by wahrwolf** [comment]
will this PR ever pass xD

@oblitum you are right though and I have no idea why I removed the %s...

**Comment by oblitum** [comment]
Added #2349 to fix this.

**Comment by skywind3000** [comment]
@w0rp , please

**Pull request merged**